pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -41,100 +41,134 @@ var client = http.Client{
 		return applyWebhookPolicy(fas.Spec.Policy.Webhook, f)
 	}
 
-	return f.Status.Replicas, false, errors.New("wrong policy type, should be one of: Buffer, Webhook")
+	return 0, false, errors.New("wrong policy type, should be one of: Buffer, Webhook")
 }
 
-func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (int32, bool, error) {
-	faReq := autoscalingv1.FleetAutoscaleReview{
-		Request: &autoscalingv1.FleetAutoscaleRequest{
-			UID:       uuid.NewUUID(),
-			Name:      f.Name,
-			Namespace: f.Namespace,
-			Status:    f.Status,
-		},
-		Response: nil,
+func applyWebhookPolicy(w *autoscalingv1.WebhookPolicy, f *agonesv1.Fleet) (replicas int32, limited bool, err error) {
+	if w == nil {
+		return 0, false, errors.New("nil WebhookPolicy passed")
 	}
-	b, err := json.Marshal(faReq)
-	urlStr := ""
-	if w.URL != nil {
-		urlStr = *w.URL
+
+	if f == nil {
+		return 0, false, errors.New("nil Fleet passed")
 	}
-	var faResp autoscalingv1.FleetAutoscaleReview
-	servicePath := ""
-	if w.Service != nil {
-		if w.Service.Path != nil {
-			servicePath = *w.Service.Path
+
+	if w.URL != nil && w.Service != nil {
+		return 0, false, errors.New("service and url cannot be used simultaneously")
+	}
+
+	var u *url.URL
+
+	if w.URL != nil {
+		if *w.URL == "" {
+			return 0, false, errors.New("URL was not provided")
 		}
+		u, err = url.ParseRequestURI(*w.URL)
 		if err != nil {
-			return f.Status.Replicas, false, err
+			return 0, false, err
+		}
+	} else {
+		if w.Service.Name == "" {
+			return 0, false, errors.New("service name was not provided")
+		}
+
+		var servicePath string
+		if w.Service.Path != nil {
+			servicePath = *w.Service.Path
 		}
 
 		if w.Service.Namespace == "" {
 			w.Service.Namespace = "default"
 		}
-		scheme := "http://"
+
+		scheme := "http"
 		if w.CABundle != nil {
-			scheme = "https://"
+			scheme = "https"
+
+			// We can have multiple fleetautoscalers with different CABundles defined,
+			// so we switch client.Transport before each POST request
+			rootCAs := x509.NewCertPool()
+			if ok := rootCAs.AppendCertsFromPEM(w.CABundle); !ok {
+				return 0, false, errors.New("no certs were appended from caBundle")
+			}
+			client.Transport = &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs: rootCAs,
+				},
+			}
 		}
-		urlStr = fmt.Sprintf("%s%s.%s.svc:8000/%s", scheme, w.Service.Name, w.Service.Namespace, servicePath)
+
+		u = buildURL(scheme, w.Service.Name, w.Service.Namespace, servicePath)
 	}
-	if urlStr == "" {
-		return f.Status.Replicas, false, errors.New("URL was not provided")
+
+	faReq := autoscalingv1.FleetAutoscaleReview{
+		Request: &autoscalingv1.FleetAutoscaleRequest{
+			UID:       uuid.NewUUID(),
+			Name:      f.Name,
+			Namespace: f.Namespace,
+			Status:    f.Status,
+		},
+		Response: nil,
 	}
 
-	u, err := url.Parse(urlStr)
+	b, err := json.Marshal(faReq)
 	if err != nil {
-		return f.Status.Replicas, false, err
+		return 0, false, err
```

Comment:
Again, why the change to 0? We have a consistent pattern across the code base of not doing this, so I'm curious why we're deviating from it at this stage?

Reply:
I didn't notice this pattern that is why I decided to make these changes.
I believe this is a common practice to return either a successful result and nil error or a default value and non-nil value. When I return 0 I explicitly show that it is not expected to process this value anymore, only an error should be handled. When I see code where both f.Status.Replicas(for example) and error are returned I am not sure whether these replicas are supposed to be handled somehow outside of this function or not. It brings some sort of ambiguity to the code, in my opinion. That is why I think it is better to return default values explicitly. According to this function, there are no places where both int and err values are used.
Why do you think we should return both values as is?

Reply:
Essentially, if you are returning an error, the value that accompany's it has no value -- since there is an error, something bad has happened, so there is no extra value in being specific and allocating a new value when a semantic one is already present.

I'm trying to find a convention wherein it should go one way or another, but no success so far. @roberthbailey you've always got thoughts on Go readability and conventions -- care to weight in and tiebreak here?

Reply:
If we are talking about allocations, we return basic types in this function - (int32, bool, error). There is no heap allocation here.

Reply:
I've usually seen what @akremsa said - you return a default value for the type (e.g. nil for a pointer, 0 for integer, "" for a string, etc) since you expect the caller to ignore the value. 

I've looked through https://golang.org/doc/effective_go.html#errors, https://blog.golang.org/error-handling-and-go, and https://github.com/golang/go/wiki/CodeReviewComments and don't see any strong guidance one way or the other. 

From those links I checked out the golang source for file.go and it uses the pattern of returning `"", err` or `0, err` depending on the function signature (e.g. https://golang.org/src/os/file.go?s=8470:8507#L151). 

So if you want me to tie-break I'd go with returning the default value for the type.

Reply:
> So if you want me to tie-break I'd go with returning the default value for the type.

Okily! In which case, I remove my objections, and resolve this conversation.